wsbroadcastserver/clientconnection.go

```diff
@@ -76,33 +86,98 @@ func (cc *ClientConnection) Compression() bool {
 	return cc.compression
 }
 
+func (cc *ClientConnection) writeBacklog(ctx context.Context, segment backlog.BacklogSegment) (backlog.BacklogSegment, error) {
+	for segment != nil {
+		select {
+		case <-ctx.Done():
+			return nil, errContextDone
+		default:
+		}
+
+		bm := &m.BroadcastMessage{
+			Messages: segment.Messages(),
+		}
+		notCompressed, compressed, err := serializeMessage(cc.clientManager, bm, !cc.compression, cc.compression)
+		if err != nil {
+			return nil, err
+		}
+
+		data := []byte{}
+		if cc.compression {
+			data = compressed.Bytes()
+		} else {
+			data = notCompressed.Bytes()
```

Comment:
I would say we shouldn't serialize both versions if we don't need to, but we expect compression to the common case, and the notCompressed serialization is presumably required to generate the compressed output, so this might be fine as-is. Still, if it isn't too difficult, it might be worth only serializing the one we need.

Reply:
I agree that the `serializeMessage` function is a little confusing. [It was already there](https://github.com/OffchainLabs/nitro/blob/e0e9a88d363184a19be92b6bf39408b28916e708/wsbroadcastserver/clientmanager.go#L257-L302) so I haven't refactored it. It doesn't actually serialize both versions, it will only do the one that we have chosen and the other returned value will be an empty slice of bytes. I think that leads to confusion and this should be refactored but I have chosen to ignore it for now to not increase the scope of this change.

I'd be happy to refactor the function in this or a future PR though if you'd like? FYI from what I can see I don't think the compressed output requires anything from the notCompressed serialization to function, so it should be easy to split into two functions.